lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -445,6 +460,70 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
                                     SN->getValue());
           RTR = F->createTranspose("transpose", NS, LTR->getShuffle());
           changed = true;
+        } else if (isa<Constant>(node->getNthInput(ArithmeticNode::LHSIdx)) &&
+                   RTR) {
+          // Build Constant' for for LHS.
+          auto *C =
+              dyn_cast<Constant>(node->getNthInput(ArithmeticNode::LHSIdx));
+          auto &T = C->getPayload();
+
+          // In order for a new Transpose node with the same shuffle as RTR to
+          // be created at the output of the Constant, the new Constant should
+          // have the same type as the input to RTR.
+          auto *NC = F->getParent()->createConstant(
+              RTR->getInput().getType(), C->getName().str() + ".transposed");
+
+          // The payload of the original Constant C has the same type as the
+          // output of RTR. In order to preserve correctness, this payload must
+          // be transposed using the inverse of the shuffle of RTR and stored
+          // into the payload of the new Constant.
+          //
+          // Another way to think of this is that we are inserting two
+          // Transposes back to back between the old Constant and the Add node.
+          // In order to sink the second Transpose below the Add node, its
+          // shuffle must be equal to that of the Transpose that is the RHS
+          // operand of the Add. In order to preserve correctness, the shuffle
+          // of the first Transpose must be the inverse of that shuffle of the
+          // second Transpose. The statement below statically computes this
+          // first Transpose.
+          T.transpose(&NC->getPayloadMutable(),
+                      invertShuffle(RTR->getShuffle()));
```

Comment:
This big snippet and a similar snippet below look almost identical. Move then into a helper function?